### PR TITLE
Fix multiple issues in `lookup_drop_cname` (supersedes #286)

### PR DESCRIPTION
**Pull request by friendlyhugo**
This PR supersedes #286 (as it cherry-picks @szenker's and @pemensik's commits).

This PR fixes a couple of issues with `lookup_drop_cname`, which cause lookups to be retained even after their corresponding CNAME records are removed. It also corrects the lookup flags comparison in `lookup_drop_cname` which never matched the `flags` value.

At the time, #286 was blocked on #551 being merged - #682, which superseded #551 was merged a couple weeks back.

Thanks to @szenker and @pemensik for their work on #286!

### CODE REVIEW

**Comment by evverx** [comment]
The third commit just says
```
avahi-core: fix flags comparison in lookup_find and lookup_drop_cname
Closes https://github.com/avahi/avahi/pull/286".
```
Can you expand on what it fixes? One use case where avahi did the wrong thing with the flags should suffice.

File: avahi-core/browse.c [hunk 1]
```diff
@@ -155,6 +155,19 @@ static AvahiSRBLookup* lookup_ref(AvahiSRBLookup *l) {
     return l;
 }
 
+static int avahi_lookup_match(AvahiSRBLookup *l, AvahiIfIndex interface, AvahiProtocol protocol, AvahiLookupFlags flags, AvahiKey *key) {
```
Comment (siteshwar) [inline avahi-core/browse.c]: Please review these [warnings](https://openscanhub.fedoraproject.org/task/78960/log/added.html) reported by OpenScanHub:
```
Error: CPPCHECK_WARNING ([CWE-398](https://cwe.mitre.org/data/definitions/398.html)): [[#def1]](https://openscanhub.fedoraproject.org/task/78960/log/added.html#def1)
avahi-0.9.rc2/avahi-core/browse.c:158: style[constParameterPointer]: Parameter 'key' can be declared as pointer to const
#  156|   }
#  157|   
#  158|-> static int avahi_lookup_match(AvahiSRBLookup *l, AvahiIfIndex interface, AvahiProtocol protocol, AvahiLookupFlags flags, AvahiKey *key) {
#  159|       if (l->interface != AVAHI_IF_UNSPEC && l->interface != interface)
#  160|           return 0;

Error: CPPCHECK_WARNING ([CWE-398](https://cwe.mitre.org/data/definitions/398.html)): [[#def2]](https://openscanhub.fedoraproject.org/task/78960/log/added.html#def2)
avahi-0.9.rc2/avahi-core/browse.c:158: style[constParameterPointer]: Parameter 'l' can be declared as pointer to const
#  156|   }
#  157|   
#  158|-> static int avahi_lookup_match(AvahiSRBLookup *l, AvahiIfIndex interface, AvahiProtocol protocol, AvahiLookupFlags flags, AvahiKey *key) {
#  159|       if (l->interface != AVAHI_IF_UNSPEC && l->interface != interface)
#  160|           return 0;
```

**Comment by friendlyhugo** [comment]
> Can you expand on what it fixes? One use case where avahi did the wrong thing with the flags should suffice.

I experienced this bug when I used CNAME records with a low TTL. When said TTL expires before the record is resolved, avahi keeps sending requests and even handles them if they are answered.

I've added a more technical explanation in the commit message.

File: avahi-core/browse.c (hunk 1 above)
Comment (friendlyhugo) [reply avahi-core/browse.c]: Fixed.

**Comment by friendlyhugo** [comment]
@evverx, I updated the commit message. WDYT?

**Comment by friendlyhugo** [comment]
I've opened #797 to track this.

**Comment by evverx** [comment]
> I've opened https://github.com/avahi/avahi/issues/797 to track this.

Let's keep discussing it there. I'll add the needs-reporter-feedback label here too.

**Comment by evverx** [comment]
While the thing where avahi keeps sending queries after CNAMEs expire is being discussed can you update the commit messages of the other commits and expand on what exactly they fix? Basically it should be possible to include that in the release notes (or whatever) to make it clear what changes. Things like

> [Fix another place with the same issue](https://github.com/avahi/avahi/pull/729/commits/c4bf836bfa043a630b8dc0aeff3dd790455f2652)
> https://github.com/friendlyhugo/avahi/commit/6d40075c8d258f01637ceb8fae07b4d64b19f64a have not noticed the same thing
is repeated multiple times. Fix that too.

aren't clear enough.

**Comment by friendlyhugo** [comment]
I updated the commit messages, but kept the original authors of #286 for proper credit.

**Comment by evverx** [comment]
> I updated the commit messages, but kept the original authors of https://github.com/avahi/avahi/pull/286 for proper credit.

Makes sense (though I don't think those commits make sense separately so they should probably be turned into one commit with a bunch of `Co-authored-by` but those details can be ironed out later)

**Comment by evverx** [comment]
Looking at the commit messages they just say what they do. Can you add stuff like "prevents avahi from sending queries when CNAMES expire" to clarify specific issues they address?

**Comment by evverx** [comment]
It turns out with your patch applied the DoS mentioned in https://github.com/avahi/avahi/issues/684 is no longer reproducible so it affects the events positively in that particular case (though I haven't looked at that closely). With that I'll add the "important" label. Either way can you update the commit messages?

**Comment by friendlyhugo** [comment]
> It turns out with your patch applied the DoS mentioned in #684 is no longer reproducible

Great, I added a "Fixes" line to the commit message.

> Either way can you update the commit messages?

I updated them and squashed the two commits from #286.

**Comment by evverx** [comment]
I'll fetch this PR next week to run some tests to see whether anything else blows up. As mentioned in https://github.com/avahi/avahi/issues/797 I'm concerned about the events related to CNAMEs and how they can affect the existing clients. avahi-browse is fine but it's not the only client consuming those events. I'll add it to the 0.9 milestone in the meantime.

**Comment by evverx** [comment]
Actually it's still possible to mess up the events and crash avahi-browse by sending another CNAME so "Fixes https://github.com/avahi/avahi/issues/684" can be removed. I'll remove the important label.

**Comment by evverx** [comment]
Looking at the following commit message
```
avahi-core: fix flags comparison in lookup_find and lookup_drop_cname

Previously avahi would only accept flags if they matched l->flags exactly.
This never actually happened, as flags was always 0 and l->flags is always
non-0 (initialized in lookup_new using transport_flags_from_domain).
```
it makes it look like `lookup_find` and `lookup_drop_cname` are equally affected and that's not accurate. Can you adjust that part?

Also why do you keep `flags != 0` in `avahi_lookup_match`?

**Comment by friendlyhugo** [comment]
> why do you keep `flags != 0` in `avahi_lookup_match`?

This is required in case `b->flags` is initialized to `0` by a client (`0` means "any transport").

Do you think we should close #797 for now? Let me know if you wanna keep working on this.

**Comment by evverx** [comment]
> This is required in case b->flags is initialized to 0 by a client (0 means "any transport").

That's correct. If `avahi-browse` used `AVAHI_LOOKUP_USE_MULTICAST` instead of `0` it wouldn't crash in https://github.com/avahi/avahi/issues/684 with that particular CNAME. What I was trying to say is that the commit message currently focuses on the thing where lookups aren't removed correctly but it doesn't say that they aren't added correctly either in some cases preventing the loop detection added in https://github.com/avahi/avahi/pull/682 from working.

> Do you think we should close https://github.com/avahi/avahi/issues/797 for now? Let me know if you wanna keep working on this

I don't think it should be closed. I think the cname test need tweaking to cover various protocols, flags and so on.

**Comment by friendlyhugo** [comment]
I updated the commit message to include the changes to `lookup_find`.

> I think the cname test need tweaking to cover various protocols, flags and so on.

Anything apart from running the test with `flags=0`? Should I add it to this branch or open a new PR?

**Comment by evverx** [comment]
> Anything apart from running the test with flags=0?

The test doesn't actually fail when it's run with `flags=0`. It just emits
```
browse.c: Failed to create SRBLookup.
browse.c: Failed to create SRBLookup.
```
and keeps going so in its current form it can't catch that. I'd need to think about how that part can be covered better.

> Should I add it to this branch or open a new PR?

I think the tests should go along with the code to make sure it actually works so it would be better to add those things here.

**Comment by friendlyhugo** [comment]
I decided to add a new test to ensure `lookup_find` functions correctly. It essentially checks whether the final lookup has been added, which, as you've mentioned, would fail before this change.

**Comment by evverx** [comment]
> It essentially checks whether the final lookup has been added

I have to say it's a really creative way to make sure that the CNAME pointing to itself isn't added a gazillion times.

I've been looking at 78eab31128479f06e30beb8c1cbf99dd921e2524 and wondering why `lookup_exists_in_path` is needed at all. If `avahi_lookup_match` can look for lookups correctly now `lookup_handle_cname` can detect loops by looking for lookups before adding them. If they already exists it's a loop. Or is there a reason that it can't be handled that way?

**Comment by evverx** [comment]
Also with the new test added the code path where `AVAHI_LOOKUPS_PER_BROWSER_MAX` is hit is no longer exercised. I think it should be covered as well by adding a bunch of different CNAMEs or something like that.

**Comment by friendlyhugo** [comment]
I encountered a weird crash on this branch. I'm looking into it, converting to draft for now.

**Comment by evverx** [comment]
What's about https://github.com/avahi/avahi/commit/78eab31128479f06e30beb8c1cbf99dd921e2524 though? It looks convoluted for no reason. I'm planning to revert https://github.com/avahi/avahi/pull/682 and then reintroduce those things one by one with proper commit messages and links to advisories.

**Comment by friendlyhugo** [comment]
> Also with the new test added the code path where AVAHI_LOOKUPS_PER_BROWSER_MAX is hit is no longer exercised.

I added a test for that.

> Or is there a reason that it can't be handled that way?

I tried implementing this idea (along with similar ones) a while back, when I worked on #682. From what I can recall, I ran into several edge cases that all of these simpler algorithms didn't handle well. For example, I added the `cname_answer_triangle_removed` test which the algorithm you suggested will incorrectly detect as a loop, which results in the entire lookup failing.

If you'd like, I can add some more comments in `lookup_exists_in_path` which will clarify it.

> I'm planning to revert #682 and then reintroduce those things one by one with proper commit messages and links to advisories.

Do you want me to reword the messages for #682?

**Comment by evverx** [comment]
Why do you think `cname_answer_triangle_removed` is correct? mDNSResponder isn't going to resolve that for example.

**Comment by evverx** [comment]
> Do you want me to reword the messages for https://github.com/avahi/avahi/pull/682

Generally the idea is to revert the PR, reopen the issues and then address them in separate PRs with proper code review and things like that (and that would include rewording the commit messages). Let's discuss this in 2026 though.

**Comment by evverx** [comment]
> Why do you think cname_answer_triangle_removed is correct? mDNSResponder isn't going to resolve that for example.

Just to be absolutely precise here mDNSResponder resolves that but it does that because (unlike avahi in this PR) it doesn't remove CNAMEs that expire. It doesn't actually follow the `X` -> `Z` -> `Y` chain.
